src/components/config_parser.cpp

```diff
@@ -0,0 +1,276 @@
+#include <fstream>
+#include <algorithm>
+
+#include "components/config_parser.hpp"
+
+POLYBAR_NS
+
+config_parser::config_parser(const logger& logger, string&& file, string&& bar)
+  : m_log(logger),
+  m_file(file_util::expand(forward<string>(file))),
+  m_barname(forward<string>(bar)) {}
+
+config::make_type config_parser::parse() {
+  m_log.info("Parsing config file: %s", m_file);
+
+  parse_file(m_file, {});
+
+  sectionmap_t sections = create_sectionmap();
+
+  if(sections.find("bar/" + m_barname) == sections.end()) {
+    throw application_error("Undefined bar: " + m_barname);
+  }
+
+  /*
+   * The first element in the files vector is always the main config file and
+   * because it has unique filenames, we can use all the elements from the
+   * second element onwards for the included list
+   */
+  file_list included(files.begin() + 1, files.end());
+  config::make_type result = config::make(m_file, m_barname);
+
+  // Cast to non-const to set sections, included and xrm
+  config& m_conf = const_cast<config&>(result);
```

Comment:
Yes, terrible practice. Also in the cpu module :smile: 

I think I will change that in future PRs and make `config` something that is constructed just like a normal struct and after that is const in the entire program

Reply:
I ave a suggestion for this.
How we have suppressed this in the cpu module.
Maybe we should make `config::make` private and add `friend class config_parser` into the `config` class then we add a `config::get` or something like that which returns the singleton. What do you think of that?

In case we want to remove the singleton, we could also store the config into `bar` and so we can access it from bar into the modules.

Reply:
I think removing the singleton would be good. The singleton provides two things: it restricts a the class to a single instance and it gives easy access to it. We only need the easy access since technically there could be multiple config instances. So moving the config into an object the code already has access to (`bar_settings` or similar) is a good idea.

Reply:
I can do that once your PR is merged if you want.

Reply:
On second though, it may be beneficial to just pass the config to wherever it's needed since it's only really needed in the modules, `controller.cpp`, `renderer.cpp`, `screen.cpp`, and `tray_manager.cpp`. Though we could probably remove it from `screen.cpp` since it only needs access to the `screenchange-reload` setting.

Reply:
Yeah that's also a possibility.